### PR TITLE
sql: correctly determine resolved ids for arrays

### LINKED ISSUE #21958: adapter: catalog consistency violation
See https://github.com/MaterializeInc/materialize/pull/21957

### PR DESCRIPTION
**Pull request by mjibson**
Explicitly described maps, arrays, and lists referenced their inner types in resolved_ids. However, on restart some types would have a named variant that was not obviously an array. For example, `int[]` would convert to `pg_catalog._int`, triggering the "named" type path instead of "array" type path. In the named path, there was no awareness of possible inner types, so the outer type would be added to resolved_ids, resulting in catalog consistency violations.

Teach the named type path to use referenced types when adding to resolved_ids so the code paths have the same results.

Fixes #21958

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Comment by mjibson** [comment]
Genuinely not sure if this is the correct approach, but it's what I came up with.

**Comment by sploiselle** [comment]
This seems to interact poorly with the way dependency tracking works for subsources--I need to dig into this more but I think that one the way into memory, subsources don't have their dependencies independently tracked in-memory, while when reading from disk, the individual statements (e.g. `CREATE SUBSOURCE`) have their dependencies handled individually.

File: src/sql/src/names.rs [hunk 1]
```diff
@@ -1440,7 +1441,17 @@ impl<'a> NameResolver<'a> {
                         (full_name, item)
                     }
                 };
+
                 self.ids.insert(item.id());
+                self.ids.extend(
```
Comment (mjibson) [inline src/sql/src/names.rs:1446]: This version only includes the types one level deep, whereas the previous version included the innermost, non container types. I have no idea which one is better.

Comment (sploiselle) [reply src/sql/src/names.rs:1446]: You're right--this should recursively find the inner elements. My oversight.

File: src/sql/src/names.rs [hunk 2]
```diff
@@ -1399,6 +1399,7 @@ impl<'a> NameResolver<'a> {
                                     .quoted()
                             ),
                         };
+                        self.ids.insert(array_item.id());
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/sql/src/names.rs:1402]: Yes because they are objects that are depended on.

**Comment by sploiselle** [comment]
`resolved_ids` is the mechanism to track dependencies, which e.g. prevents dropping items that still have dependents. This hasn't been a problem for us because the types in question are all system objects that can't be dropped anyway.

**Comment by sploiselle** [comment]
> Then we should probably include both the outer and all inner types? List of map of record of X should include all those as dependencies?

Yes. The complexity with this is just that we poorly handle subsource dependencies at the moment.

File: src/sql/src/names.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/sql/src/names.rs:1402]: Sorry this is a very late comment. What is a list or map outer id? My understanding is that the outer id of an array are the types of the form `_int4`, `_char`, etc. However, I didn't think that those existed for maps and lists.

File: src/sql/src/names.rs (hunk 1 above)
Comment (jkosh44) [reply src/sql/src/names.rs:1446]: Again, sorry for the late comment. It's not clear to me why we'd want to recurse here. For example if we have the following view definitions:

```SQL
CREATE VIEW v1 AS SELECT 1;
CREATE VIEW v2 AS SELECT * FROM v1;
CREATE VIEW v3 AS SELECT * FROM v2;

SELECT * FROM v3;
```

When planning the `SELECT` statement above, only `v3` will appear in the list of resolved IDs. We don't recurse and include `v1` and `v2`. So why do we treat types differently?

**Pull request closed without merging**